File: S/907-sum-of-subarray-minimums/main.go
package main

func sumSubarrayMins2(arr []int) int {
	M := 1000000007
	left, right := make([]int, len(arr)), make([]int, len(arr))
	for i := 0; i < len(arr); i++ {
		base := arr[i]
		cnt := 0
		for j := i; j < len(arr); j++ {
			if arr[j] >= base {
				cnt++
			} else {
				break
			}
		}
		left[i] = cnt
	}

	for i := len(arr) - 1; i >= 0; i-- {
		base := arr[i]
		cnt := 1
		for j := i - 1; j >= 0; j-- {
			if arr[j] > base {
				cnt++
			} else {
				break
			}
		}
		right[i] = cnt
	}
	// fmt.Println(left, right)

	sum := 0
	for i := 0; i < len(arr); i++ {
		sum += (left[i] * right[i] * arr[i])
		sum %= M
	}
	return sum
}

File: S/907-sum-of-subarray-minimums/main.rs
struct Solution {}

impl Solution {
  pub fn sum_subarray_mins(arr: Vec<i32>) -> i32 {
    let mut stack: Vec<(i32, usize)> = Vec::new();
    let mut left: Vec<usize> = vec![arr.len(); arr.len()];
    let mut right: Vec<usize> = vec![arr.len(); arr.len()];

    arr.iter().enumerate().for_each(|(idx, &v)| {
      while stack.len() > 0 && v <= stack[stack.len() - 1].0 {
        let (t, offset) = stack.pop().unwrap();
        left[offset] = idx;
      }
      stack.push((v, idx));
    });
    stack.clear();
    arr.iter().rev().enumerate().for_each(|(idx, &v)| {
      while stack.len() > 0 && v < stack[stack.len() - 1].0 {
        let (t, offset) = stack.pop().unwrap();
        right[arr.len() - 1 - offset] = idx;
      }
      stack.push((v, idx));
    });

    let mut ret: i64 = 0;
    (0..arr.len()).for_each(|idx| {
      ret +=
        (arr[idx] as i64 * (left[idx] - idx) as i64 * (idx + 1 + right[idx] - arr.len()) as i64)
          % 1000000007;
      ret %= 1000000007;
    });
    ret as i32
  }
}
